p2p/protocol.go

```diff
@@ -304,19 +316,72 @@ func (c *conn) handleNewBlockHashes(ctx context.Context, msg ethp2p.Msg) error {
 
 	c.counter.WithLabelValues(packet.Name(), c.node.URLv4(), c.name).Add(float64(len(packet)))
 
-	hashes := make([]common.Hash, 0, len(packet))
-	for _, hash := range packet {
-		hashes = append(hashes, hash.Hash)
-		if err := c.getBlockData(hash.Hash); err != nil {
+	// Collect unique hashes for database write.
+	uniqueHashes := make([]common.Hash, 0, len(packet))
+
+	for _, entry := range packet {
+		hash := entry.Hash
+
+		// Check if we've seen the hash and remove old entries
+		if c.hasSeenBlockHash(hash) {
+			c.logger.Info().Str("hash", hash.Hex()).Msg("Skipping duplicate block hash")
+			continue
+		}
+
+		// Attempt to fetch block data first
+		if err := c.getBlockData(hash); err != nil {
 			return err
 		}
+
+		// Now that we've successfully fetched, record the new block hash
+		c.addBlockHash(hash)
+		uniqueHashes = append(uniqueHashes, hash)
 	}
 
-	c.db.WriteBlockHashes(ctx, c.node, hashes, tfs)
+	// Write only unique hashes to the database.
+	if len(uniqueHashes) > 0 {
+		c.db.WriteBlockHashes(ctx, c.node, uniqueHashes, tfs)
+	}
 
 	return nil
 }
 
+// addBlockHash adds a new block hash with a timestamp to the blockHashes list.
+func (c *conn) addBlockHash(hash common.Hash) {
+	now := time.Now()
+
+	// Add the new block hash entry to the list.
+	c.blockHashes.PushBack(BlockHashEntry{
+		hash: hash,
+		time: now,
+	})
+}
+
+// Helper method to check if a block hash is already in blockHashes.
+func (c *conn) hasSeenBlockHash(hash common.Hash) bool {
+	now := time.Now()
+	found := false
+
+	for e := c.blockHashes.Front(); e != nil; {
+		next := e.Next()
+		entry := e.Value.(BlockHashEntry)
+
+		// Remove entries older than blockHashTTL.
+		if now.Sub(entry.time) > blockHashTTL {
+			c.blockHashes.Remove(e)
+		} else {
+			// Check if the hash matches.
+			if entry.hash.Cmp(hash) == 0 {
+				found = true
+			}
+		}
+
+		e = next
+	}
```

Comment:
```suggestion
// Helper method to check if a block hash is already in blockHashes.
func (c *conn) hasSeenBlockHash(hash common.Hash) bool {
	now := time.Now()

	for e := c.blockHashes.Front(); e != nil; e = e.Next() {
		entry := e.Value.(BlockHashEntry)

		// Check if the hash matches. We can short circuit here because there will
		// be block hashes that we haven't seen before, which will make a full
		// iteration of the blockHashes linked list.
		if entry.hash.Cmp(hash) == 0 {
			return true
		}

		// Remove entries older than blockHashTTL.
		if now.Sub(entry.time) > blockHashTTL {
			c.blockHashes.Remove(e)
		}
	}

	return false
}
```
don't commit this suggestion, just copy paste it in for formatting issues